### ISSUE TITLE
xfomers related: llvm assert failed(core dumped)

### ISSUE THREAD

**Issue by Howe-Young** [siliconflow/onediff#424]
我在from onediff.infer_compiler import oneflow_compile的时候报如下错误：
`python: /home/ci-user/manylinux-cache-dir/release/cu118/build/oneflow/ir/llvm_monorepo-src/llvm/include/llvm/Support/CommandLine.h:864: void llvm::cl::parser<DataType>::addLiteralOption(llvm::StringRef, const DT&, llvm::StringRef) [with DT = llvm::FunctionPass* (*)(); DataType = llvm::FunctionPass* (*)()]: Assertion `findOption(Name) == Values.size() && "Option already exists!"' failed.
Aborted (core dumped)`

包配置如下：
Package                   Version                 Location
------------------------- ----------------------- ------------------------------------------------------------
absl-py                   2.0.0
accelerate                0.25.0
aiohttp                   3.8.6
aiosignal                 1.3.1
albumentations            1.3.0
altair                    5.0.1
antlr4-python3-runtime    4.8
appdirs                   1.4.4
async-timeout             4.0.3
attrs                     23.1.0
blinker                   1.6.2
braceexpand               0.1.7
cachetools                5.3.1
certifi                   2023.7.22
charset-normalizer        3.3.1
click                     8.1.3
clip                      1.0                     
cmake                     3.27.7
coloredlogs               15.0.1
contourpy                 1.2.0
cuda-python               12.3.0
cycler                    0.12.1
decorator                 5.1.1
diffusers                 0.24.0
docker-pycreds            0.4.0
einops                    0.3.0
filelock                  3.13.1
flatbuffers               23.5.26
fonttools                 4.44.0
frozenlist                1.4.0
fsspec                    2023.10.0
ftfy                      6.1.1
future                    0.18.3
gitdb                     4.0.10
GitPython                 3.1.37
google-auth               2.23.4
google-auth-oauthlib      1.0.0
grpcio                    1.59.2
huggingface-hub           0.19.4
humanfriendly             10.0
idna                      3.4
imageio                   2.9.0
imageio-ffmpeg            0.4.2
importlib-metadata        6.8.0
importlib-resources       6.1.0
invisible-watermark       0.1.5
Jinja2                    3.1.2
joblib                    1.3.2
jsonschema                4.19.2
jsonschema-specifications 2023.7.1
kiwisolver                1.4.5
kornia                    0.6.0
lazy_loader               0.3
lit                       17.0.4
Markdown                  3.4.3
markdown-it-py            3.0.0
MarkupSafe                2.1.3
matplotlib                3.7.1
mdurl                     0.1.2
mpmath                    1.3.0
mssapi                    1.1
multidict                 6.0.4
mypy-extensions           1.0.0
networkx                  3.1
numpy                     1.24.4
nvidia-cublas-cu11        11.10.3.66
nvidia-cublas-cu12        12.1.3.1
nvidia-cuda-cupti-cu12    12.1.105
nvidia-cuda-nvrtc-cu11    11.7.99
nvidia-cuda-nvrtc-cu12    12.1.105
nvidia-cuda-runtime-cu11  11.7.99
nvidia-cuda-runtime-cu12  12.1.105
nvidia-cudnn-cu11         8.5.0.96
nvidia-cudnn-cu12         8.9.2.26
nvidia-cufft-cu12         11.0.2.54
nvidia-curand-cu12        10.3.2.106
nvidia-cusolver-cu12      11.4.5.107
nvidia-cusparse-cu12      12.1.0.106
nvidia-nccl-cu12          2.18.1
nvidia-nvjitlink-cu12     12.3.52
nvidia-nvtx-cu12          12.1.105
oauthlib                  3.2.2
omegaconf                 2.1.1
onediff                   0.11.0.dev0             /home/yh/gitDir/onediff/src
oneflow                   0.9.1.dev20231213+cu118
onefx                     0.0.3
onnx                      1.14.0
onnxconverter-common      1.13.0
onnxmltools               1.11.2
onnxruntime-gpu           1.12.0
onnxsim                   0.4.35
open-clip-torch           2.0.2
opencv-python             4.6.0.66
opencv-python-headless    4.8.1.78
packaging                 23.2
pandas                    2.0.2
pathtools                 0.1.2
Pillow                    10.1.0
pip                       21.2.4
protobuf                  3.20.3
psutil                    5.9.5
pyarrow                   14.0.0
pyasn1                    0.5.0
pyasn1-modules            0.3.0
pydeck                    0.8.1b0
pyDeprecate               0.3.1
Pygments                  2.16.1
Pympler                   1.0.1
pyparsing                 3.1.1
pyre-extensions           0.0.29
python-cat                0.0.11
python-dateutil           2.8.2
pytorch-lightning         1.4.2
pytz                      2023.3.post1
PyWavelets                1.4.1
PyYAML                    6.0.1
qudida                    0.0.4
referencing               0.30.2
regex                     2023.10.3
requests                  2.31.0
requests-oauthlib         1.3.1
rich                      13.6.0
rpds-py                   0.7.1
rsa                       4.9
safetensors               0.4.1
scikit-image              0.20.0
scikit-learn              1.2.2
scipy                     1.9.1
semver                    3.0.2
sentry-sdk                1.34.0
setproctitle              1.3.2
setuptools                57.4.0
six                       1.16.0
skl2onnx                  1.14.1
smmap                     5.0.1
StableSR                  0.0.1                  
streamlit                 1.12.1
streamlit-drawable-canvas 0.8.0
sympy                     1.12
taming-transformers       0.0.1                   
tensorboard               2.13.0
tensorboard-data-server   0.7.2
tensorrt                  9.0.1.post11.dev4
test-tube                 0.7.5
threadpoolctl             3.2.0
tifffile                  2023.9.26
tokenizers                0.15.0
toml                      0.10.2
toolz                     0.12.0
torch                     2.0.0+cu117
torchmetrics              0.6.0
torchvision               0.15.1+cu117
tornado                   6.3.3
tqdm                      4.65.0
transformers              4.36.0
triton                    2.0.0
typing_extensions         4.8.0
typing-inspect            0.9.0
tzdata                    2023.3
tzlocal                   5.2
urllib3                   2.0.7
validators                0.20.0
wandb                     0.15.12
watchdog                  3.0.0
wcwidth                   0.2.8
webdataset                0.2.5
Werkzeug                  3.0.1
wheel                     0.37.0
xformers                  0.0.19
yarl                      1.9.2
zipp                      3.17.0

**Comment by olokojoh** [comment]
same issue

**Comment by Howe-Young** [comment]
> 我在from onediff.infer_compiler import oneflow_compile的时候报如下错误： `python: /home/ci-user/manylinux-cache-dir/release/cu118/build/oneflow/ir/llvm_monorepo-src/llvm/include/llvm/Support/CommandLine.h:864: void llvm::cl::parser<DataType>::addLiteralOption(llvm::StringRef, const DT&, llvm::StringRef) [with DT = llvm::FunctionPass* (*)(); DataType = llvm::FunctionPass* (*)()]: Assertion `findOption(Name) == Values.size() && "Option already exists!"' failed. Aborted (core dumped)`
> 
> 包配置如下： Package Version Location
> 
> absl-py 2.0.0 accelerate 0.25.0 aiohttp 3.8.6 aiosignal 1.3.1 albumentations 1.3.0 altair 5.0.1 antlr4-python3-runtime 4.8 appdirs 1.4.4 async-timeout 4.0.3 attrs 23.1.0 blinker 1.6.2 braceexpand 0.1.7 cachetools 5.3.1 certifi 2023.7.22 charset-normalizer 3.3.1 click 8.1.3 clip 1.0 cmake 3.27.7 coloredlogs 15.0.1 contourpy 1.2.0 cuda-python 12.3.0 cycler 0.12.1 decorator 5.1.1 diffusers 0.24.0 docker-pycreds 0.4.0 einops 0.3.0 filelock 3.13.1 flatbuffers 23.5.26 fonttools 4.44.0 frozenlist 1.4.0 fsspec 2023.10.0 ftfy 6.1.1 future 0.18.3 gitdb 4.0.10 GitPython 3.1.37 google-auth 2.23.4 google-auth-oauthlib 1.0.0 grpcio 1.59.2 huggingface-hub 0.19.4 humanfriendly 10.0 idna 3.4 imageio 2.9.0 imageio-ffmpeg 0.4.2 importlib-metadata 6.8.0 importlib-resources 6.1.0 invisible-watermark 0.1.5 Jinja2 3.1.2 joblib 1.3.2 jsonschema 4.19.2 jsonschema-specifications 2023.7.1 kiwisolver 1.4.5 kornia 0.6.0 lazy_loader 0.3 lit 17.0.4 Markdown 3.4.3 markdown-it-py 3.0.0 MarkupSafe 2.1.3 matplotlib 3.7.1 mdurl 0.1.2 mpmath 1.3.0 mssapi 1.1 multidict 6.0.4 mypy-extensions 1.0.0 networkx 3.1 numpy 1.24.4 nvidia-cublas-cu11 11.10.3.66 nvidia-cublas-cu12 12.1.3.1 nvidia-cuda-cupti-cu12 12.1.105 nvidia-cuda-nvrtc-cu11 11.7.99 nvidia-cuda-nvrtc-cu12 12.1.105 nvidia-cuda-runtime-cu11 11.7.99 nvidia-cuda-runtime-cu12 12.1.105 nvidia-cudnn-cu11 8.5.0.96 nvidia-cudnn-cu12 8.9.2.26 nvidia-cufft-cu12 11.0.2.54 nvidia-curand-cu12 10.3.2.106 nvidia-cusolver-cu12 11.4.5.107 nvidia-cusparse-cu12 12.1.0.106 nvidia-nccl-cu12 2.18.1 nvidia-nvjitlink-cu12 12.3.52 nvidia-nvtx-cu12 12.1.105 oauthlib 3.2.2 omegaconf 2.1.1 onediff 0.11.0.dev0 /home/yh/gitDir/onediff/src oneflow 0.9.1.dev20231213+cu118 onefx 0.0.3 onnx 1.14.0 onnxconverter-common 1.13.0 onnxmltools 1.11.2 onnxruntime-gpu 1.12.0 onnxsim 0.4.35 open-clip-torch 2.0.2 opencv-python 4.6.0.66 opencv-python-headless 4.8.1.78 packaging 23.2 pandas 2.0.2 pathtools 0.1.2 Pillow 10.1.0 pip 21.2.4 protobuf 3.20.3 psutil 5.9.5 pyarrow 14.0.0 pyasn1 0.5.0 pyasn1-modules 0.3.0 pydeck 0.8.1b0 pyDeprecate 0.3.1 Pygments 2.16.1 Pympler 1.0.1 pyparsing 3.1.1 pyre-extensions 0.0.29 python-cat 0.0.11 python-dateutil 2.8.2 pytorch-lightning 1.4.2 pytz 2023.3.post1 PyWavelets 1.4.1 PyYAML 6.0.1 qudida 0.0.4 referencing 0.30.2 regex 2023.10.3 requests 2.31.0 requests-oauthlib 1.3.1 rich 13.6.0 rpds-py 0.7.1 rsa 4.9 safetensors 0.4.1 scikit-image 0.20.0 scikit-learn 1.2.2 scipy 1.9.1 semver 3.0.2 sentry-sdk 1.34.0 setproctitle 1.3.2 setuptools 57.4.0 six 1.16.0 skl2onnx 1.14.1 smmap 5.0.1 StableSR 0.0.1 streamlit 1.12.1 streamlit-drawable-canvas 0.8.0 sympy 1.12 taming-transformers 0.0.1 tensorboard 2.13.0 tensorboard-data-server 0.7.2 tensorrt 9.0.1.post11.dev4 test-tube 0.7.5 threadpoolctl 3.2.0 tifffile 2023.9.26 tokenizers 0.15.0 toml 0.10.2 toolz 0.12.0 torch 2.0.0+cu117 torchmetrics 0.6.0 torchvision 0.15.1+cu117 tornado 6.3.3 tqdm 4.65.0 transformers 4.36.0 triton 2.0.0 typing_extensions 4.8.0 typing-inspect 0.9.0 tzdata 2023.3 tzlocal 5.2 urllib3 2.0.7 validators 0.20.0 wandb 0.15.12 watchdog 3.0.0 wcwidth 0.2.8 webdataset 0.2.5 Werkzeug 3.0.1 wheel 0.37.0 xformers 0.0.19 yarl 1.9.2 zipp 3.17.0

解决了，卸载了xfomers就行了，是因为oneflow和xformers不能同时安装吗，底层有冲突？我编译安装的xformers也不行

**Comment by strint** [comment]
> 我在from onediff.infer_compiler import oneflow_compile的时候报如下错误： `python: /home/ci-user/manylinux-cache-dir/release/cu118/build/oneflow/ir/llvm_monorepo-src/llvm/include/llvm/Support/CommandLine.h:864: void llvm::cl::parser<DataType>::addLiteralOption(llvm::StringRef, const DT&, llvm::StringRef) [with DT = llvm::FunctionPass* (*)(); DataType = llvm::FunctionPass* (*)()]: Assertion `findOption(Name) == Values.size() && "Option already exists!"' failed. Aborted (core dumped)`

@jackalcooper 

安装了 xformers，看起来是和 llvm 里面一个配置有冲突？

**Comment by jackalcooper** [comment]
系统里面有其他llvm库会被加载吗

**Comment by zhangvia** [comment]
可以考虑先import oneflow，后import torch xformer之类的库。

**Comment by jackalcooper** [comment]
Note
- older version of `xformers` has this problem (for instance 0.0.19)
- newer version of xf works (for instance 0.0.23.post1+cu118)

**Comment by strint** [comment]
> Note
> 
> * older version of `xformers` has this problem (for instance 0.0.19)
> * newer version of xf works (for instance 0.0.23.post1+cu118)

So we recommend the user to upgrade the xformers version to avoid thi problem.

or import oneflow before torch?

**Comment by jackalcooper** [comment]
> So we recommend the user to upgrade the xformers version to avoid this problem.
> 
> or import oneflow before torch?

Yes.

**Comment by Howe-Young** [comment]
> 可以考虑先import oneflow，后import torch xformer之类的库。

这种导入不会出错，但是推理一样会出错

**Comment by strint** [comment]
> > 可以考虑先import oneflow，后import torch xformer之类的库。
> 
> 这种导入不会出错，但是推理一样会出错

可以试下这个：
newer version of xf works (for instance 0.0.23.post1+cu118)

**Comment by jackalcooper** [comment]
newer oneflow build will resolve this issue

**Comment by InuyashaLee** [comment]
同时使用xformers和onediff报错：
![1](https://github.com/siliconflow/onediff/assets/114284586/5ebb9c25-63e2-4c38-affb-0c88c5b3ed73)
我的代码：
![2](https://github.com/siliconflow/onediff/assets/114284586/08296a76-9730-4580-8d42-c37b299893c8)
相关包：
![3](https://github.com/siliconflow/onediff/assets/114284586/6cbb00db-c7c9-478d-9291-2c85a16a56d0)
cuda：
![4](https://github.com/siliconflow/onediff/assets/114284586/0409abdf-3276-44ce-9d0b-c17ec9e3ec25)
但是先oneflow_compile推理一遍后再开启xformer不报错（不过速度和显存占用未见提升）
![5](https://github.com/siliconflow/onediff/assets/114284586/7625ba12-9569-405b-8e90-7e7352fa5fbc)
请问这些现象是正常的吗？另外，先oneflow_compile推理一遍后再开启xformer这种用法是必须的吗？

**Comment by jackalcooper** [comment]
> 请问这些现象是正常的吗？另外，先oneflow_compile推理一遍后再开启xformer这种用法是必须的吗？

- oneflow_compile 已经包含 xformer 类似的优化，所以用 oneflow_compile 优化模型的时候不需要使用 xformer。
- 本 issue 之前讨论的是系统里面安装过 xformer 造成的兼容问题，这个问题已经解决了。

**Comment by strint** [comment]
> 先oneflow_compile推理一遍后再开启xformer这种用法是必须的吗？

请也参考下这里的回答：https://github.com/siliconflow/onediff/issues/467

@InuyashaLee
